http-webhook.md

```diff
@@ -0,0 +1,333 @@
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][RFC2119].
+
+### 1.2. Relation to HTTP
+
+This specification prescribes rules constraining the use and handling of
+specific [HTTP methods][RFC7231-Section-4] and headers.
+
+This specification also applies equivalently to HTTP/2 ([RFC7540][RFC7540]),
+which is compatible with HTTP 1.1 semantics.
+
+## 2. Delivering notifications
+
+### 2.1. Delivery request
+
+Notifications are delivered using a HTTP request. The response indicates the
+resulting status of the delivery.
+
+HTTP-over-TLS (HTTPS) [RFC2818][RFC2818] MUST be used for the connection.
+
+The HTTP method for the delivery request MUST be [POST][POST].
+
+The [`Content-Type`][Content-Type] header MUST be carried and the request MUST
+carry a notification payload of the given content type. Requests without
+payloads, e.g. where the notification is entirely expressed in HTTP headers,
+are not permitted by this specification.
+
+This specification does not further constrain the content of the notification,
+and it also does not prescribe the [HTTP target resource][RFC7230-Section-5-1]
+that is used for delivery.
+
+If the delivery target supports and requires [Abuse Protection](#4-abuse-protection),
+the delivery request MUST include the `Origin` header. The `Origin` header value
+is a DNS name expression that identifies the sending system.
+
+### 2.2. Delivery response
+
+The delivery response MAY contain a payload providing detail status
+information in the case of handling errors. This specification does not attempt
+to define such a payload.
+
+The response MUST NOT use any of the [3xx HTTP Redirect status codes][3xx] and
+the client MUST NOT follow any such redirection.
+
+If a delivery target has been retired, but the HTTP site still exists, the
+site SHOULD return a [410 "Gone"][410] status code and the sender SHOULD refrain
+from sending any further notifications.
+
+If the delivery target is unable to process the request due to exceeding a
+request rate limit, it SHOULD return a [429 Too Many Requests][429] status code
+and MUST include the [`Retry-After`][Retry-After] header. The sender MUST
+observe the value of the Retry-After header and refrain from sending further
+requests until the indicated time.
+
+If the delivery has been accepted and processed, and if the response carries a
+payload with processing details, the response MUST have the [200 OK][200]
+status code. In this case, the response MUST carry a
+[`Content-Type`][Content-Type] header.
+
+If the delivery has been accepted and processed, but carries no payload, the
+response MUST have the [204 No Content][204] status code.
+
+If the delivery has been accepted, but has not yet been processed or if the
+processing status is unknown, the response MUST have the [202 Accepted][202]
+status code.
+
+## 3. Authorization
+
+The delivery request MUST use one of the following two methods, both of which
```

Comment:
"following two methods" --> "following two authorization methods" ?

More importantly, is this "MUST" requirement for delivery request and (in next sentence) for delivery target, with practically one option only, really necessary? This alone would eliminate (make non-compliant) many current OSS FaaS solutions, and any notification system which uses k8s+Istio RBAC for implementing security aspect.

Reply:
App/workload logic is agnostic of the security aspect. Aspect like security is requested declaratively and enforced by Istio, where side-car container injected into protected workloads intercepts incoming and outgoing traffic to uniformly, typically across the k8s cluster, apply cross-cutting concerns like security (identity, authorization, mutual TLS), traffic management (weighted routing), tracing, monitoring. See https://istio.io/docs/concepts/security/rbac.html

Reply:
I looked at this in the spec context again and now stand by the existing wording. The request MUST use either of the two options shown. If you have some magic interceptor thing, then that magic interceptor thng must make the request in aggregate compliant with the rule here before it gets to the target. This spec aims to define interop between a sender and a target that use open standards without coupling to some shared security framework.

I see that ISTIO provides a model to establish mutual TLS auth, but I do not understand how that established identity surfaces in an HTTP request other than through whatever proprietary context mechanism a web server gives you. If it's completely transparent also for the receiving endpoint we're effectively talking about unauthenticated HTTP plaintext at both ends that's aided by interceptor magic, and I don't know how to build a secure standard for the endpoints on top of that.